### PR TITLE
DRAFT: add microk8s functions, config - todo: tests, fix config

### PR DESCRIPTION
**Pull request by aliwatters**
Fixes: #5327 (in progress!)

**Description**

Support microk8s loading images -- see: https://microk8s.io/docs/registry-images

**DRAFT PR**

Unsure on this section in `loadimages.go`
```
// loadImagesInMicrok8sNodes loads artifact images into every node of a microk8s cluster.
func (r *SkaffoldRunner) loadImagesInMicrok8sNodes(ctx context.Context, out io.Writer, k8sCluster string, artifacts []build.Artifact) error {
	color.Default.Fprintln(out, "Loading images into microk8s cluster nodes...")
	return r.loadImages(ctx, out, artifacts, func(tag string) *exec.Cmd {
		// Note: needs multipass installed and a vm of name 'microk8s-vm' -- how to ensure?

		// 1. build the image
		// docker save image-name | multipass transfer - microk8s-vm:image-name.tar
		exec.CommandContext(ctx, "docker", "save", "image-name", "|", "multipass", "transfer", "-", "microk8s-vm:image-name.tar")

		// 2. load into microk8s
		// microk8s ctr image import image-name.tar
		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "image-name.tar", tag) // note --cluster isn't valid
	})
}
```

## Additional TODO:

More tests.

### CODE REVIEW

**Comment by aliwatters** [comment]
gofmt errors -- used to have this setup to run magically.

```
RUN hack/gofmt.sh
Gofmt errors in files:
./pkg/skaffold/runner/load_images.go
./pkg/skaffold/runner/deploy.go
./pkg/skaffold/config/util.go
diff -u ./pkg/skaffold/runner/load_images.go.orig ./pkg/skaffold/runner/load_images.go
--- ./pkg/skaffold/runner/load_images.go.orig	2021-02-04 07:38:19.406621418 -0800
+++ ./pkg/skaffold/runner/load_images.go	2021-02-04 07:38:19.406621418 -0800
@@ -53,7 +53,7 @@
 	color.Default.Fprintln(out, "Loading images into microk8s cluster nodes...")
 	return r.loadImages(ctx, out, artifacts, func(tag string) *exec.Cmd {
 		// first pass -- not confident -- import is looking for a file
-		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid 
+		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid
 	})
 }
 
diff -u ./pkg/skaffold/runner/deploy.go.orig ./pkg/skaffold/runner/deploy.go
--- ./pkg/skaffold/runner/deploy.go.orig	2021-02-04 07:38:19.418622338 -0800
+++ ./pkg/skaffold/runner/deploy.go	2021-02-04 07:38:19.418622338 -0800
@@ -134,7 +134,7 @@
 			return fmt.Errorf("loading images into microk8s nodes: %w", err)
 		}
 	}
-	
+
 	return nil
 }
 
diff -u ./pkg/skaffold/config/util.go.orig ./pkg/skaffold/config/util.go
--- ./pkg/skaffold/config/util.go.orig	2021-02-04 07:38:19.966664344 -0800
+++ ./pkg/skaffold/config/util.go	2021-02-04 07:38:19.966664344 -0800
@@ -228,7 +228,6 @@
 	k3dDisableLoad := cfg.K3dDisableLoad != nil && *cfg.K3dDisableLoad
 	microk8sDisableLoad := cfg.Microk8sDisableLoad != nil && *cfg.Microk8sDisableLoad
 
-
 	// load images for local kind/k3d/microk8s cluster unless explicitly disabled
 	loadImages := local && ((isKindCluster && !kindDisableLoad) || (isK3dCluster && !k3dDisableLoad) || (isMicrok8sCluster && !microk8sDisableLoad))
 
@@ -300,9 +299,9 @@
 
 // IsMicrok8sCluster checks that the given `kubeContext` is talking to `microk8s`.
 func IsMicrok8sCluster(kubeContext string) bool {
-  return strings.HasPrefix(kubeContext, "microk8s-")
+	return strings.HasPrefix(kubeContext, "microk8s-")
 }
-	
+
 // Microk8sClusterName returns the internal name of a microk8s cluster.
 func Microk8sClusterName(clusterName string) string {
 	if strings.HasPrefix(clusterName, "microk8s-") {
FAILED hack/gofmt.sh
RUN hack/golangci-lint.sh
pkg/skaffold/runner/deploy.go:137: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)
	
pkg/skaffold/runner/load_images.go:56: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)
		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid 
pkg/skaffold/config/util.go:231: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)

FAILED hack/golangci-lint.sh
make: *** [Makefile:111: test] Error 1
RUN hack/gofmt.sh
Gofmt errors in files:
./pkg/skaffold/runner/load_images.go
./pkg/skaffold/runner/deploy.go
./pkg/skaffold/config/util.go
diff -u ./pkg/skaffold/runner/load_images.go.orig ./pkg/skaffold/runner/load_images.go
--- ./pkg/skaffold/runner/load_images.go.orig	2021-02-04 07:38:19.406621418 -0800
+++ ./pkg/skaffold/runner/load_images.go	2021-02-04 07:38:19.406621418 -0800
@@ -53,7 +53,7 @@
 	color.Default.Fprintln(out, "Loading images into microk8s cluster nodes...")
 	return r.loadImages(ctx, out, artifacts, func(tag string) *exec.Cmd {
 		// first pass -- not confident -- import is looking for a file
-		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid 
+		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid
 	})
 }
 
diff -u ./pkg/skaffold/runner/deploy.go.orig ./pkg/skaffold/runner/deploy.go
--- ./pkg/skaffold/runner/deploy.go.orig	2021-02-04 07:38:19.418622338 -0800
+++ ./pkg/skaffold/runner/deploy.go	2021-02-04 07:38:19.418622338 -0800
@@ -134,7 +134,7 @@
 			return fmt.Errorf("loading images into microk8s nodes: %w", err)
 		}
 	}
-	
+
 	return nil
 }
 
diff -u ./pkg/skaffold/config/util.go.orig ./pkg/skaffold/config/util.go
--- ./pkg/skaffold/config/util.go.orig	2021-02-04 07:38:19.966664344 -0800
+++ ./pkg/skaffold/config/util.go	2021-02-04 07:38:19.966664344 -0800
@@ -228,7 +228,6 @@
 	k3dDisableLoad := cfg.K3dDisableLoad != nil && *cfg.K3dDisableLoad
 	microk8sDisableLoad := cfg.Microk8sDisableLoad != nil && *cfg.Microk8sDisableLoad
 
-
 	// load images for local kind/k3d/microk8s cluster unless explicitly disabled
 	loadImages := local && ((isKindCluster && !kindDisableLoad) || (isK3dCluster && !k3dDisableLoad) || (isMicrok8sCluster && !microk8sDisableLoad))
 
@@ -300,9 +299,9 @@
 
 // IsMicrok8sCluster checks that the given `kubeContext` is talking to `microk8s`.
 func IsMicrok8sCluster(kubeContext string) bool {
-  return strings.HasPrefix(kubeContext, "microk8s-")
+	return strings.HasPrefix(kubeContext, "microk8s-")
 }
-	
+
 // Microk8sClusterName returns the internal name of a microk8s cluster.
 func Microk8sClusterName(clusterName string) string {
 	if strings.HasPrefix(clusterName, "microk8s-") {
FAILED hack/gofmt.sh
RUN hack/golangci-lint.sh
pkg/skaffold/runner/deploy.go:137: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)
	
pkg/skaffold/runner/load_images.go:56: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)
		return exec.CommandContext(ctx, "microk8s", "ctr", "image", "import", "--cluster", k8sCluster, tag) // note --cluster isn't valid 
pkg/skaffold/config/util.go:231: File is not `goimports`-ed with -local github.com/GoogleContainerTools/skaffold (goimports)

FAILED hack/golangci-lint.sh
make: *** [Makefile:111: test] Error 1
```

**Comment by MarlonGamez** [comment]
Thanks for taking a stab at this @aliwatters ! I think we'll have to use a couple commands in `loadImagesInMicrok8sNodes` to get the job done.

What I suggested on #5327 was to do this
```bash
docker save image-name > image-name.tar
multipass transfer image-name.tar microk8s-vm:image-name.tar
microk8s ctr image import image-name.tar
```

But I think we can translate that to something simpler to avoid writing a file to the host's disk
```bash
docker save image-name | multipass transfer - microk8s-vm:image-name.tar
microk8s ctr image import image-name.tar
```

If we can execute this in code then I think things should work fine. The only concern is having the requirement that the user has multipass installed

**Comment by aliwatters** [comment]
Newbie question -- how do I run just a single test file? -- `make test` runs through the entire suite, `go test <test_file>` fails with missing env vars.

**Comment by aliwatters** [comment]
Having a few issues with the commands as suggested;

```
ali@stinky:~/git/skaffold (5327-load-images-to-microk8s)$ multipass version
multipass  1.6.0-rc.274+g9c637461
multipassd 1.6.0-rc.274+g9c637461

ali@stinky:~/git/dkc-multi-k8s (feature-skaffold-minimal)$ docker save aliwatters/dkc-multi-server | multipass transfer - microk8s-vm:image-name.tar
transfer failed: instance "microk8s-vm" does not exist

ali@stinky:~/git/dkc-multi-k8s (feature-skaffold-minimal)$ docker save aliwatters/dkc-multi-server | multipass transfer - microk8s-cluster:image-name.tar
transfer failed: instance "microk8s-cluster" does not exist


ali@stinky:~/git/dkc-multi-k8s (feature-skaffold-minimal)$ docker save aliwatters/dkc-multi-server | multipass transfer - microk8s:image-name.tar
transfer failed: instance "microk8s" does not exist

ali@stinky:~/git/skaffold (5327-load-images-to-microk8s)$ more ~/.kube/config 
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJ <SNIP> NBVEUtLS0tLQo=
    server: https://127.0.0.1:16443
  name: microk8s-cluster
contexts:
- context:
    cluster: microk8s-cluster
    user: admin
  name: microk8s
current-context: microk8s
kind: Config
preferences: {}
users:
- name: admin
  user:
    token: VDlqa3BnWlR4Q3UzaHhpQnFvamlkbDdBRk9aMzdBcDkvMFhZT21ad1VqND0K
```

This command does work for me;

```
ali@stinky:~/git/dkc-multi-k8s (feature-skaffold-minimal)$ docker save aliwatters/dkc-multi-server > image.tar

ali@stinky:~/git/dkc-multi-k8s (feature-skaffold-minimal)$ microk8s ctr image import image.tar
unpacking docker.io/aliwatters/dkc-multi-server:1dc48c4 (sha256:5de85e79d42b0057ecfd5553431c9800b291777a92cbfc295968b7e8d38f8b43)...done
unpacking docker.io/aliwatters/dkc-multi-server:7b9b7d87ab81483cdb9b1227780d68dec815fe271aea2f9fe3a0e65090d90909 (sha256:5de85e79d42b0057ecfd5553431c9800b291777a92cbfc295968b7e8d38f8b43)...done
unpacking docker.io/aliwatters/dkc-multi-server:90d1283 (sha256:5de85e79d42b0057ecfd5553431c9800b291777a92cbfc295968b7e8d38f8b43)...done
unpacking docker.io/aliwatters/dkc-multi-server:e9d99b8 (sha256:5de85e79d42b0057ecfd5553431c9800b291777a92cbfc295968b7e8d38f8b43)...done
unpacking docker.io/aliwatters/dkc-multi-server:f450a07 (sha256:5de85e79d42b0057ecfd5553431c9800b291777a92cbfc295968b7e8d38f8b43)...done
```

Does multipass need to be involved? Please advise.

**Comment by aliwatters** [comment]
Found my missing vm -- I hadn't created it in multipass.

```
$ multipass launch --name microk8s-vm

$ multipass list
Name                    State             IPv4             Image
microk8s-vm             Running           10.117.155.213   Ubuntu 20.04 LTS
```

Commands work as expected now.

**Comment by aliwatters** [comment]
Updated command to use `multipass` -- not yet working as hoped.

Failing tests:

```
=== Failed Tests ===
/pkg/skaffold/config/TestGetCluster
--- FAIL: TestGetCluster (0.00s)
/pkg/skaffold/config/TestGetCluster/microk8s_with_disable-load=true
    util_test.go:383: config.Cluster differ (-got, +want):   config.Cluster{
          	Local:      true,
        - 	PushImages: false,
        + 	PushImages: true,
        - 	LoadImages: true,
        + 	LoadImages: false,
          }
    util_test.go:386: config.Cluster differ (-got, +want):   config.Cluster{
          	Local:      true,
        - 	PushImages: false,
        + 	PushImages: true,
        - 	LoadImages: true,
        + 	LoadImages: false,
          }
    --- FAIL: TestGetCluster/microk8s_with_disable-load=true (0.00s)
```

Can't see why.

**Comment by MarlonGamez** [comment]
@aliwatters running `make test` runs this command `go test -count=1 -race -short -timeout=90s ./pkg/skaffold/... ./cmd/... ./hack/... ./pkg/webhook/...`

Usually to just run tests on a single package/directory I'll do this:
`go test -count=1 -race -short -timeout=90s ./path/to/package/... `

File: pkg/skaffold/runner/load_images.go [hunk 1]
```diff
@@ -48,6 +48,22 @@ func (r *SkaffoldRunner) loadImagesInK3dNodes(ctx context.Context, out io.Writer
...
 
+// loadImagesInMicrok8sNodes loads artifact images into every node of a microk8s cluster.
+func (r *SkaffoldRunner) loadImagesInMicrok8sNodes(ctx context.Context, out io.Writer, k8sCluster string, artifacts []build.Artifact) error {
+	color.Default.Fprintln(out, "Loading images into microk8s cluster nodes...")
+	return r.loadImages(ctx, out, artifacts, func(tag string) *exec.Cmd {
+		// Note: needs multipass installed and a vm of name 'microk8s-vm' -- how to ensure?
+
+		// 1. build the image
+		// docker save image-name | multipass transfer - microk8s-vm:image-name.tar
+		exec.CommandContext(ctx, "docker", "save", "image-name", "|", "multipass", "transfer", "-", "microk8s-vm:image-name.tar")
```
Comment (tejal29) [inline pkg/skaffold/runner/load_images.go:59]: the `image_name` should correspond to the image built by docker right?
shdn't this be the image name in the `tag` field without the tag?

File: pkg/skaffold/runner/deploy.go [hunk 2]
```diff
@@ -126,6 +126,15 @@ func (r *SkaffoldRunner) loadImagesIntoCluster(ctx context.Context, out io.Write
 		}
 	}
 
+	if config.IsMicrok8sCluster(r.runCtx.GetKubeContext()) {
+		microk8sCluster := config.Microk8sClusterName(currentContext.Cluster)
```
Comment (tejal29) [inline pkg/skaffold/runner/deploy.go:130]: `config.IsMicrok8sCluster` function can be moved to `deploy` package and defined as a local  function as it does not rely on any utilities in `config`

**Comment by MarlonGamez** [comment]
hey @aliwatters , just wanted to do a quick pulse check :) Let me know if you need help with anything or have any questions!

**Comment by nkubala** [comment]
@aliwatters this one is getting pretty stale, and i'm wondering if we should keep it open or close it out for now. let us know if you're still thinking about trying to push this through and someone from the team can provide some assistance!

**Comment by nkubala** [comment]
@aliwatters i'm going to close this for now. please feel free and reopen if you ever pick it back up, and the team can help you get it merged!

**Pull request closed without merging**